mezo-org/mezod
# Migrate mezo-node-4 under Helm chart

References: https://linear.app/thesis-co/issue/ENG-368/migrate-nodes-living-on-mezo-staging-to-the-validator-kit-helm-chart





### Introduction

This PR moves `mezo-node-4` under the official Helm chart from the Validator Kit. The deployment is defined using code for the [Helmfile](https://github.com/helmfile/helmfile).

### Helmfile

Helmfile is similar to Terraform. Under the good it uses Helm binary, so it must be installed in the system. The file containing the code can be named `helmfile.yaml`.

Helmfile installation:

```shell
brew install helmfile
# helmfile uses helm and helm-diff plugin, so to install necessary plugins run:
helmfile init
```

Useful commands:

- `helmfile list` - list releases from `helmfile.yaml`
- `helmfile template` - run `helm template` for given release
- `helmfile diff` - show the plan of changes (similar to `terraform plan`)
- `helmfile apply` - show diff and apply changes

Keep in mind, that `helmfile apply` doesn't ask user for confirmation like Terraform does. This is the most important difference :)
To make `helmfile` interactive you have to add a flag `--interactive` or `-i` or set environment variable `HELMFILE_INTERACTIVE=true`.

There are some other useful flags. For example:

- `--selector`/`-l` - select release using labels like, `name`
- `--context 5` - show diff with only 5 lines above or below the change



### Changes



### What was done?

The migration required one additional step since the previous Pod mounted PVC in `$MEZOD_HOME/data`, but the new one uses `$MEZOD_HOME`. We need to move files to have the right structure.

1. Remove StatefulSet for `mezo-node-4` and it's Service:
```
kubectl delete -f mezo-node-4.yaml
```
2. Set `maintenanceMode` to `true`. The Pod will start without application, but the structure of disk will be created.
3. Run `helmfile apply -l name=mezo-node-4 -i`.
4. Scale down Statefulset: `kubectl scale --replicas=0 statefulset/mezo-node-4 -n default`
5. Apply this Pod:
```
apiVersion: v1
kind: Pod
metadata:
  name: mezod-pvc-migrator
  namespace: default
spec:
  containers:
    - name: app
      image: ubuntu:latest
      command:
        - /bin/sh
        - -c
        - 'trap "exit" TERM; while true; do sleep 1; done'
      volumeMounts:
      - name: data
        mountPath: /storage
  volumes:
  - name: data
    persistentVolumeClaim:
      claimName: data-mezo-node-4-0
```
It mounts the PVC in `/storage` folder.
6. Exec into Pod's shell: `kubectl exec -it mezod-pvc-migrator -- /bin/bash`
7. Move files:
```
mv /storage/*.db /storage/data
mv /storage/priv_validator_state.json /storage/data/
mv /storage/snapshots /storage/data/
```
6. Exit and remove the Pod
7. Set `maintenanceMode` to `false`
8. Run `helmfile apply -l name=mezo-node-4 -i`

It should work! Good luck!


### Testing



Trust me, I'm an engineer.

---

### Author's checklist


### Reviewer's checklist
